tests/model_integration/test_model.py

```diff
@@ -68,3 +68,32 @@ def test_stream_add_multiple(selected_model):
     lm += ""
     *_, last_lm = lm
     assert str(last_lm) in ["item1", "item2"]
+
+
+def test_associativity(selected_model):
+    prompt = "pi = "
+    grammar = gen("number", regex=r"\d")
+    engine = selected_model.engine
+
+    with patch.object(engine, "get_next_token", side_effect=engine.get_next_token) as get_next_token_1:
+        _ = selected_model + (prompt + grammar)
+    prompt_tokens_1 = get_next_token_1.call_args.kwargs["token_ids"]
+
+    with patch.object(engine, "get_next_token", side_effect=engine.get_next_token) as get_next_token_2:
+        _ = (selected_model + prompt) + grammar
+    prompt_tokens_2 = get_next_token_2.call_args.kwargs["token_ids"]
+
+    # Main assertion: the prompt tokens should be the same
+    assert prompt_tokens_1 == prompt_tokens_2
+
+    # Further assert that the tokenization matches the expected tokenization
+    expected_prompt_tokens = engine.tokenizer.encode(prompt.encode())
+    if (
+        engine.tokenizer.bos_token is not None
+        and expected_prompt_tokens[:1] != [engine.tokenizer.bos_token_id]
+    ):
+        expected_prompt_tokens = [engine.tokenizer.bos_token_id] + expected_prompt_tokens
+        expected_prompt_tokens = engine.tokenizer.recode(expected_prompt_tokens)
```

Comment:
@riedgar-ms would love your input on this test, especially regarding recode.

Reply:
I wouldn't expect `recode()` to have any effect here. Surely a bos_token should always recode to a bos_token no matter what follows it? If not, I would expect there to be some spectacular jail breaks.

Shouldn't you always have `expected_prompt_tokens == prompt_tokens_1 == prompt_tokens_2`? Or at most just have to worry about slicing the bos_token off the front of the actual tokens, based on whether or not the engine has a bos_token?

Reply:
Try running the test without the recode here -- some llamacpp models seem to need the recode to fix the token(s) immediately following the BOS token. May have something to do with whitespace preceding the first bit of text

And then we have to allow for the tokens that get passed to the model to have one or more tokens sliced off the end because of token healing